File: app/src/main/java/com/simon/harmonichackernews/utils/StoryPreviewImageMemoryCache.java
package com.simon.harmonichackernews.utils;

import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.util.LruCache;

import androidx.annotation.Nullable;

public final class StoryPreviewImageMemoryCache {

    private static final int MAX_ENTRIES = 48;
    private static final LruCache<String, Drawable.ConstantState> CACHE =
            new LruCache<>(MAX_ENTRIES);
    private static final LruCache<String, Integer> TINT_CACHE =
            new LruCache<>(MAX_ENTRIES);

    private StoryPreviewImageMemoryCache() {
    }

    public static void put(int storyId, String imageUrl, @Nullable Drawable drawable) {
        if (storyId <= 0 || TextUtils.isEmpty(imageUrl) || drawable == null) {
            return;
        }

        Drawable.ConstantState constantState = drawable.getConstantState();
        if (constantState != null) {
            synchronized (CACHE) {
                CACHE.put(getKey(storyId, imageUrl), constantState);
            }
        }
    }

    @Nullable
    public static Drawable get(int storyId, String imageUrl) {
        if (storyId <= 0 || TextUtils.isEmpty(imageUrl)) {
            return null;
        }

        Drawable.ConstantState constantState;
        synchronized (CACHE) {
            constantState = CACHE.get(getKey(storyId, imageUrl));
        }
        return constantState == null ? null : constantState.newDrawable();
    }

    public static void putTintColor(int storyId, String imageUrl, int baseColor, int tintColor) {
        if (storyId <= 0 || TextUtils.isEmpty(imageUrl)) {
            return;
        }

        synchronized (TINT_CACHE) {
            TINT_CACHE.put(getTintKey(storyId, imageUrl, baseColor), tintColor);
        }
    }

    @Nullable
    public static Integer getTintColor(int storyId, String imageUrl, int baseColor) {
        if (storyId <= 0 || TextUtils.isEmpty(imageUrl)) {
            return null;
        }

        synchronized (TINT_CACHE) {
            return TINT_CACHE.get(getTintKey(storyId, imageUrl, baseColor));
        }
    }

    private static String getKey(int storyId, String imageUrl) {
        return storyId + ":" + imageUrl;
    }

    private static String getTintKey(int storyId, String imageUrl, int baseColor) {
        return getKey(storyId, imageUrl) + ":" + baseColor;
    }
}

File: app/src/main/java/com/simon/harmonichackernews/utils/StoryPreviewImageMemoryCache.kt
package com.simon.harmonichackernews.utils

import android.util.LruCache

object StoryPreviewImageMemoryCache {
    private const val MAX_ENTRIES = 48
    private const val TINT_CACHE_VERSION = "3"
    private val tintCache = LruCache<String, Int>(MAX_ENTRIES)

    fun putTintColor(storyId: Int, imageUrl: String?, baseColor: Int, tintColor: Int) {
        if (storyId <= 0 || imageUrl.isNullOrEmpty()) return
        synchronized(tintCache) {
            tintCache.put(getTintKey(storyId, imageUrl, baseColor), tintColor)
        }
    }

    fun getTintColor(storyId: Int, imageUrl: String?, baseColor: Int): Int? {
        if (storyId <= 0 || imageUrl.isNullOrEmpty()) return null
        return synchronized(tintCache) {
            tintCache.get(getTintKey(storyId, imageUrl, baseColor))
        }
    }

    fun clearTintColors() {
        synchronized(tintCache) {
            tintCache.evictAll()
        }
    }

    private fun getKey(storyId: Int, imageUrl: String): String = "$storyId:$imageUrl"

    private fun getTintKey(storyId: Int, imageUrl: String, baseColor: Int): String =
        "${getKey(storyId, imageUrl)}:$TINT_CACHE_VERSION:$baseColor"
}
